src/Capability/Discovery/CachedDiscoverer.php

```diff
@@ -0,0 +1,149 @@
+<?php
+
+/*
+ * This file is part of the official PHP MCP SDK.
+ *
+ * A collaboration between Symfony and the PHP Foundation.
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Mcp\Capability\Discovery;
+
+use Psr\Log\LoggerInterface;
+use Psr\SimpleCache\CacheInterface;
+
+/**
+ * Cached decorator for the Discoverer class.
+ *
+ * This decorator caches the results of file system operations and reflection
+ * to improve performance when discovery is called multiple times.
+ *
+ * @author Xentixar <xentixar@gmail.com>
+ */
+class CachedDiscoverer
+{
+    private const CACHE_PREFIX = 'mcp_discovery_';
+    private const CACHE_TTL = 3600; // 1 hour default TTL
+
+    public function __construct(
+        private readonly Discoverer $discoverer,
+        private readonly CacheInterface $cache,
+        private readonly LoggerInterface $logger,
+        private readonly int $cacheTtl = self::CACHE_TTL,
+    ) {
+    }
+
+    /**
+     * Discover MCP elements in the specified directories with caching.
+     *
+     * @param string        $basePath    the base path for resolving directories
+     * @param array<string> $directories list of directories (relative to base path) to scan
+     * @param array<string> $excludeDirs list of directories (relative to base path) to exclude from the scan
+     */
+    public function discover(string $basePath, array $directories, array $excludeDirs = []): DiscoveryState
+    {
+        $cacheKey = $this->generateCacheKey($basePath, $directories, $excludeDirs);
+
+        // Check if we have cached results
+        $cachedResult = $this->cache->get($cacheKey);
+        if (null !== $cachedResult) {
+            $this->logger->debug('Using cached discovery results', [
+                'cache_key' => $cacheKey,
+                'base_path' => $basePath,
+                'directories' => $directories,
+            ]);
+
+            // Restore the discovery state from cache
+            return $this->restoreDiscoveryStateFromCache($cachedResult);
```

Comment:
if it's possible to cache the object instead of the array, we could skip this method here and make it simpler:
```suggestion
            return $cachedResult;
```